arch: arm: aarch32: Fix arch_cpu_idle interrupt masking and add memory barriers

stephanosio opened the pull request:
```
arch: arm: aarch32: Fix arch_cpu_idle interrupt masking

The current AArch32 `arch_cpu_idle` implementation enables interrupt
before executing the WFI instruction, and this has the side effect of
allowing interruption and thereby calling wake-up notification
functions before the CPU enters sleep.

This commit fixes the problem described above by ensuring that
interrupt is disabled when the WFI instruction is executed and
re-enabled only after the processor wakes up.

For ARMv6-M, ARMv8-M Baseline and ARM-R, the PRIMASK (ARM-M)/
CPSR.I (ARM-R) is used to lock interrupts and therefore it is not
necessary to do anything before executing the WFI instruction.

For ARMv7-M and ARMv8-M Mainline, the BASEPRI is used to lock
interrupts and the PRIMASK is always cleared in non-interrupt context;
therefore, it is necessary to set the PRIMASK to mask interrupts,
before clearing the BASEPRI to configure wake-up interrupt priority to
the lowest.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>
```

```
This commit adds the required memory barriers to the `arch_cpu_idle`
function in order to ensure proper idle operation in all cases.

1. Add ISB after setting BASEPRI to ensure that the new wake-up
  interrupt priority is visible to the WFI instruction.

2. Add DSB before WFI to ensure that all memory transactions are
  completed before going to sleep.

3. Add ISB after CPSIE to ensure that the pending wake-up interrupt
  is serviced immediately.

Co-authored-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>
Signed-off-by: Stephanos Ioannidis <root@stephanos.io>
```

Addresses #22078

ioannisg committed:
arch: arm: aarch32: correct documentation of arch_cpu_atomic_idle

z_CpuIdleInit has been renamed to z_arm_cpu_idle_init, so
we need to correct that function's name in the documentation
of arch_cpu_atomic_idle.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

stephanosio committed:
arch: arm: aarch32: Fix arch_cpu_idle interrupt masking

The current AArch32 `arch_cpu_idle` implementation enables interrupt
before executing the WFI instruction, and this has the side effect of
allowing interruption and thereby calling wake-up notification
functions before the CPU enters sleep.

This commit fixes the problem described above by ensuring that
interrupt is disabled when the WFI instruction is executed and
re-enabled only after the processor wakes up.

For ARMv6-M, ARMv8-M Baseline and ARM-R, the PRIMASK (ARM-M)/
CPSR.I (ARM-R) is used to lock interrupts and therefore it is not
necessary to do anything before executing the WFI instruction.

For ARMv7-M and ARMv8-M Mainline, the BASEPRI is used to lock
interrupts and the PRIMASK is always cleared in non-interrupt context;
therefore, it is necessary to set the PRIMASK to mask interrupts,
before clearing the BASEPRI to configure wake-up interrupt priority to
the lowest.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio commented:
Note that this may or may not be a problem depending on the microarchitecture implementation.

For instance, for the M profile cores (at least the ones that are currently available), _"`CPSIE` that performs interrupt enabling is not guaranteed to be self synchronizing, with the architecture guaranteeing the recognition of the change only following an ISB." <sup>[1]</sup>_

While that may be the case, the new implementation provides a fundamentally safe approach (regardless of the microarchitecture implementation), and, IMO, should be used.

In re: #23436, since we are adding ISB after CPSIE/MSR to BASEPRI, any pending interrupt is guaranteed to be executed and we **must** use this PR's approach.

<sup>[1]</sup> ARM AN 321: see "CPS and MSR instructions" under "3.3 Architectural Requirements"

stephanosio commented on arch/arm/core/aarch32/cpu_idle.S:
```diff
@@ -62,19 +62,23 @@ SECTION_FUNC(TEXT, arch_cpu_idle)
-#endif /* CONFIG_ARMV6_M_ARMV8_M_BASELINE */
+#endif /* CONFIG_ARMV7_M_ARMV8_M_MAINLINE */
 
 	wfi
```
In re: #23436, we must `DSB` before `WFI`.

stephanosio commented on arch/arm/core/aarch32/cpu_idle.S:
```diff
@@ -62,19 +62,23 @@ SECTION_FUNC(TEXT, arch_cpu_idle)
 	wfi
 
+	/* Clear PRIMASK to service any pending interrupt */
+	cpsie i
```
In re: #23436, we should `ISB` here to ensure that any pending interrupt (in particular, the one that caused a wake-up) is executed immediately.

stephanosio committed:
arch: arm: aarch32: Add memory barriers to arch_cpu_idle

This commit adds the required memory barriers to the `arch_cpu_idle`
function in order to ensure proper idle operation in all cases.

1. Add ISB after setting BASEPRI to ensure that the new wake-up
  interrupt priority is visible to the WFI instruction.

2. Add DSB before WFI to ensure that all memory transactions are
  completed before going to sleep.

3. Add ISB after CPSIE to ensure that the pending wake-up interrupt
  is serviced immediately.

Co-authored-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>
Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

ioannisg commented on arch/arm/core/aarch32/cpu_idle.S:
```diff
@@ -62,19 +62,36 @@ SECTION_FUNC(TEXT, arch_cpu_idle)
 #endif /* CONFIG_ARMV6_M_ARMV8_M_BASELINE */
 #endif /* CONFIG_TRACING */
 
```
nit: can you fix indentation on line 59?

stephanosio replied:
Not related to this patch, but will add a commit here to fix that.

stephanosio replied:
Added a commit for overall whitespace clean-up.

ioannisg replied:
thanks

ioannisg commented on arch/arm/core/aarch32/cpu_idle.S:
```diff
@@ -62,19 +62,36 @@ SECTION_FUNC(TEXT, arch_cpu_idle)
-	cpsie i
-#elif defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
-	/* clear BASEPRI so wfi is awakened by incoming interrupts */
+#if defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
```
@stephanosio for the non-Cortex-M Mainline case, I do not feel it is OK to rely on the called to disable interrupts; we should do our own CPSID here, to be independent.

It is unlikely that irq_lock is going to change, but I don't want a dependence on what arch_irq_lock is doing.

stephanosio replied:
One of the reasons for keeping interrupts locked until the processor wakes up, in case of invocation from the idle thread, is to ensure that `_isr_wrapper` is not executed before going to sleep.

This is necessary because the idle thread sets `_kernel.idle`, and the `_isr_wrapper` checks this value and calls `z_sys_power_save_idle_exit` if set to a non-zero value. Locking interrupts before setting `_kernel.idle` and unlocking after waking up ensures that `z_sys_power_save_idle_exit` is executed only after the processor wakes up.

When `k_cpu_idle` (`arch_cpu_idle`) is called from a context other than the idle thread, as long as the caller does not set `_kernel.idle`, there should not be any problem. If the caller sets `_kernel.idle`, it would be the caller's responsibility to lock interrupts before doing so (adding `CPSID` here does not help in this case).

ioannisg replied:
Ah, right I see the point.

The problem is: right now there is an inconsistency between Baseline and Mainline.
What you describe here:

>When k_cpu_idle (arch_cpu_idle) is called from a context other than the idle thread, as long as the caller does not set _kernel.idle, there should not be any problem. If the caller sets _kernel.idle, it would be the caller's responsibility to lock interrupts before doing so (adding CPSID here does not help in this case).

...also holds for Mainline, but here you do call CPSID.

It is hard for the reader to understand why CPSID is not required for Baseline. We need to either add it, or then add a comment about why it is only added for v7-M

stephanosio replied:
> ...also holds for Mainline, but here you do call CPSID.

The CPSID for Mainline is there because Zephyr uses BASEPRI for interrupt (un)locking, and this value dictates the wake-up interrupt priority.

Since we must set BASEPRI to the lowest priority level to ensure proper wake-up, and this will certainly cause interrupts to be serviced unless we set `PRIMASK`, we must `CPSID`.

In summary, this is only an artifact of the current Zephyr Cortex-M Mainline interrupt management implementation, nothing more.

pabigot replied:
This seems related to my notes when trying to understand zephyr system power management recently:

> `k_cpu_idle()` ... must unlock the `arch_irq_lock()` taken by the idle thread when it's active, even though that API is documented as callable only when there is no lock.  Apparently that really means no spinlock.

Or maybe no `irq_lock`.  Since the difference between them is rather subtle (#12537).

I haven't looked closely at this PR yet to determine if that confusion remains, but if it does it'd be nice to eliminate it.

ioannisg replied:
>The CPSID for Mainline is there because Zephyr uses BASEPRI for interrupt (un)locking, and this value dictates the wake-up interrupt priority.
Since we must set BASEPRI to the lowest priority level to ensure proper wake-up, and this will certainly cause interrupts to be serviced unless we set PRIMASK, we must CPSID.
In summary, this is only an artifact of the current Zephyr Cortex-M Mainline interrupt management implementation, nothing more.

Sure, I only want this to be documented for ease of reading ( as I suggested above, but maybe not to clear :) )

stephanosio replied:
> Sure, I only want this to be documented for ease of reading ( as I suggested above, but maybe not to clear :) )

@ioannisg I will try to think of a nice way to describe this without going into too much details, as that would probably confuse the reader even more.

stephanosio replied:
Added explanation for non-BASEPRI case.

ioannisg replied:
thanks

ioannisg commented on arch/arm/core/aarch32/cpu_idle.S:
```diff
@@ -62,19 +62,36 @@ SECTION_FUNC(TEXT, arch_cpu_idle)
+
+	/*
+	 * Set wake-up interrupt priority to the lowest and synchronise to
+	 * ensure that this is visible for the WFI instruction execution.
```
visible to**, perhaps?

stephanosio replied:
Sure, I will fix that.

ioannisg commented on arch/arm/core/aarch32/cpu_idle.S:
```diff
@@ -62,19 +62,36 @@ SECTION_FUNC(TEXT, arch_cpu_idle)
+
+#endif /* CONFIG_ARMV7_M_ARMV8_M_MAINLINE */
+
+	/* Ensure that all memory transactions are completed */
```
Nope, that's DMB. the text for DSB should probably be "Ensure that no more instructions are executed before all memory transactions are completed", or something similar.

stephanosio replied:
well, "ensure ... is completed" would mean that it will block until "..." is completed, which has the same meaning as "no more instructions are executed before ..."

I will add a more verbose description, in case that is not clear enough.

ioannisg commented on arch/arm/core/aarch32/cpu_idle.S:
```diff
@@ -62,19 +62,36 @@ SECTION_FUNC(TEXT, arch_cpu_idle)
 	wfi
 
+	/*
+	 * Clear PRIMASK and flush instruction buffer to immediately service
```
This comment is also not covering 100% percent of the cases, right? It definitely correspond to the majority of the cases, but there is a corner case where WFI does not put the core into low-power because the wake-up conditions are satisfied from the beginning. That's in case the function is called with some IRQ pending...

stephanosio replied:
Sure, `WFI` will not sleep if there is a pending interrupt; though, the pending interrupt could still be considered a "wake-up interrupt" as it satisfied one of the "wake-up conditions" for the wake-up/not sleeping.

stephanosio replied:
The comment could have stated "Clear PRIMASK and flush instruction buffer to immediately service any pending interrupt," but I decided to emphasise/remind that the interrupt being serviced here is (should be) the one that satisfied the wake-up condition.

ioannisg replied:
OK

ioannisg committed:
doc: reference: kernel: remove inconsistent comment for k_cpu_idle

k_cpu_idle() API does not need to be invoked with interrupts
unlocked; it is actually invoked with interrupts locked by
the kernel CPU idling mechanism. In most architectures, the
function is, actually, un-locking interrupts itself. We need
to remove this comment from the documentation of the CPU
idle API. We add a note about the un-locking of interrupts.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented on doc/reference/kernel/other/cpu_idle.rst:
```diff
@@ -25,8 +25,8 @@ Making the CPU idle
 ===================
 
 Making the CPU idle is simple: call the k_cpu_idle() API. The CPU will stop
-executing instructions until an event occurs. Make sure interrupts are not
```
This did not seem to be accurate any more @andrewboie @andyross to confirm :)

andyross replied:
This is... not our clearest interface.  The way k_cpu_idle() works is that it's:

1. Called with interrupts **disabled**
2. Expected not to return until an interrupt has occurred and been handled (though this isn't a firm requirement)
3. Returns with interrupts **enabled**.  No, I don't know why either.  No, it doesn't honor a irq_lock() key passed in from the caller, it just unmasks them unconditionally.

So removing that sentence is correct, but we should probably rewrite this section more fully.

ioannisg replied:
So, then, let's remove the sentence and complete the description as you propose :)

ioannisg replied:
Btw - this is how it works in the idle thread.

ioannisg replied:
@andyross i've done the following
- Added your point (3) in the description of k_cpu_idle example in this section. IMO, this section (paragraph) is only there to give an example of how k_cpu_idle() may work as a standalone API, i.e. _outside_ the context of the Zephyr idle thread. Your point (1) is specific to how the function is used in the idle thread.
- Extended the documentation of k_cpu_idle and arch_cpu_idle APIs with points (3) and (2), respectively. 
Let me know your thoughts about it.

@andrewboie ^^ as well; appreciate some feedback, as this touches an all-ARCH API

andyross replied:
This looks great to me.  I think the only argument against documenting this is that it makes it incrementally harder to fix.  I mean what we'd really **want** is for us to coordinate changing all the architectures to have simpler semantics (e.g. have it not muck with the interrupt mask state and just act like any other function that expects to be called with interrupts masked).

ioannisg replied:
> This looks great to me. I think the only argument against documenting this is that it makes it incrementally harder to fix. I mean what we'd really **want** is for us to coordinate changing all the architectures to have simpler semantics (e.g. have it not muck with the interrupt mask state and just act like any other function that expects to be called with interrupts masked).

@andyross then I will remove the added documentation in the function specification, to make it easier for you to change this later.

I now only added a note that "in some architectures" the function is unconditionally enabling IRQs upon return; that's needed to raise awareness of what's really happening and does not stop Kernel maintainers from changing the semantics, IMO.

stephanosio committed:
arch: arm: aarch32: Fix whitespaces in cpu_idle.S

This commit fixes whitespaces in cpu_idle.S.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

ioannisg commented on arch/arm/core/aarch32/cpu_idle.S:
```diff
@@ -44,86 +44,119 @@ GTEXT(arch_cpu_atomic_idle)
-	msr BASEPRI, r0
+#if defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
+	/*
+	 * PRIMASK is always cleared on ARMv7-M and ARMv8-M Mainline, and
```
Maybe add the reasoning here, too, as you did in line 86.
E.g "PRIMASK is ....... Mainline (not used for interrupt locking), and ..."

stephanosio replied:
Done

ioannisg commented on arch/arm/core/aarch32/cpu_idle.S:
```diff
@@ -44,86 +44,119 @@ GTEXT(arch_cpu_atomic_idle)
+#endif /* CONFIG_ARMV7_M_ARMV8_M_MAINLINE */
+
+	/*
+	 * Wait for all memory transactions to complete before entering low
```
Yeap, descriptions is now much better.
Although an un-experience ARM developer may still confuse this with DMB.
Strictly speaking DSB means "No instruction in program order after this instruction can execute until this instruction completes."
But I am OK with the current text.

ioannisg committed:
kernel: cpu idle: add info about IRQ unlocking in k_cpu_idle

We add a note in k_cpu_idle() documentation, stressing that for
certain architectures. the function unmasks interrupts
unconditionally before returning. In the
documentation of the architecture-specific API (arch_cpu_idle)
we describe the expected behavior with regards to the wake-up
event.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

stephanosio commented:
Rebased
